### PR TITLE
Ensure pkg_resources on EL8 nodes for Obal

### PR DESCRIPTION
**Pull request by ehelms**

### CODE REVIEW

File: puppet/modules/slave/manifests/packaging/rpm.pp [hunk 1]
```diff
@@ -33,6 +33,14 @@
   $yaml = if $facts['os']['release']['major'] == '7' { 'python36-PyYAML' } else { 'python3-pyyaml' }
   ensure_packages(['python3', $yaml])
 
+  # Required by Obsah / Obal
+  if $is_el8 {
+    package { 'pkg_resources':
+      ensure   => installed,
+      provider => 'pip3.11',
```
Comment (ehelms) [inline puppet/modules/slave/manifests/packaging/rpm.pp:40]: Hrmm, not sure how to handle this.

File: puppet/modules/slave/manifests/packaging/rpm.pp [hunk 2]
```diff
@@ -33,6 +33,14 @@
   $yaml = if $facts['os']['release']['major'] == '7' { 'python36-PyYAML' } else { 'python3-pyyaml' }
   ensure_packages(['python3', $yaml])
 
+  # Required by Obsah / Obal
+  if $is_el8 {
+    package { 'pkg_resources':
```
Comment (ekohl) [inline puppet/modules/slave/manifests/packaging/rpm.pp:38]: This is part of setuptools. Can you install `python3-setuptools` or `python3.11-setuptools`? instead? Given we install `python3-pyyaml` above I'd think `python3-setuptools` is the right package.

**Review by ekohl — CHANGES_REQUESTED** [review]
Question is: should `obal` depend on setuptools? Or be migrated: https://setuptools.pypa.io/en/latest/pkg_resources.html states:
> Use of pkg_resources is deprecated in favor of [importlib.resources](https://docs.python.org/3/library/importlib.resources.html#module-importlib.resources), [importlib.metadata](https://docs.python.org/3/library/importlib.metadata.html#module-importlib.metadata) and their backports ([importlib_resources](https://pypi.org/project/importlib_resources), [importlib_metadata](https://pypi.org/project/importlib_metadata)). Some useful APIs are also provided by [packaging](https://pypi.org/project/packaging) (e.g. requirements and version parsing). Users should refrain from new usage of pkg_resources and should work to port to importlib-based solutions.

**Comment by ehelms** [comment]
> Question is: should `obal` depend on setuptools? Or be migrated: https://setuptools.pypa.io/en/latest/pkg_resources.html states:
> 
> >

I would assume there is another way and replace the use of pkg_resources:

https://github.com/search?q=repo%3Atheforeman%2Fobsah%20pkg_resources&type=code

File: puppet/modules/slave/manifests/packaging/rpm.pp (hunk 2 above)
Comment (ehelms) [reply puppet/modules/slave/manifests/packaging/rpm.pp]: Ahh yes, that did it! Thanks!

**Comment by ekohl** [comment]
https://github.com/theforeman/obsah/pull/37

File: puppet/modules/slave/manifests/packaging/rpm.pp [hunk 3]
```diff
@@ -33,6 +33,13 @@
   $yaml = if $facts['os']['release']['major'] == '7' { 'python36-PyYAML' } else { 'python3-pyyaml' }
   ensure_packages(['python3', $yaml])
 
+  # Required by Obsah / Obal
+  if $is_el8 {
+    package { 'python3.11-setuptools':
```
[the remark this answers is not in the document]
Comment (ehelms) [reply puppet/modules/slave/manifests/packaging/rpm.pp]: Yes. Let me know if this update helps with the clarity.

File: puppet/modules/slave/manifests/packaging/rpm.pp [hunk 4]
```diff
@@ -30,9 +31,16 @@
...
-  $yaml = if $facts['os']['release']['major'] == '7' { 'python36-PyYAML' } else { 'python3-pyyaml' }
+  $yaml = if $facts['os']['release']['major'] == '7' { 'python36-PyYAML' } else { "${ansible_python_version}-pyyaml" }
   ensure_packages(['python3', $yaml])
 
+  # Required by Obsah / Obal
+  if $is_el8 {
+    package { "${ansible_python_version}-setuptools":
+      ensure   => installed,
+    }
+  }
```
Comment (ekohl) [inline puppet/modules/slave/manifests/packaging/rpm.pp:42]: How about
```suggestion
  $obal_packages = if $facts['os']['release']['major'] == '7' { ['python36-PyYAML'] } else { ["${ansible_python_version}-pyyaml", "${ansible_python_version}-setuptools"] }
  ensure_packages(['python3'] + $obal_packages)
  }
```

Perhaps add some newlines.

**Review by ekohl — COMMENTED** [review]
It does help, but how about taking it a step further?

File: puppet/modules/slave/manifests/packaging/rpm.pp [hunk 5]
```diff
@@ -7,6 +7,7 @@
   Optional[String] $koji_certificate = undef,
 ) {
   $is_el7 = $facts['os']['release']['major'] == '7'
+  $ansible_python_version = if $facts['os']['release']['major'] == '8' { 'python3.11' } else { 'python3' }
```
Comment (evgeni) [inline puppet/modules/slave/manifests/packaging/rpm.pp:10]: I took the liberty to also make this el9 compatible where ansible (again) uses default python.

**Comment by evgeni** [comment]
It feels wrong to ack/merge this, as I touched the code, but we should really get it in as builds are starting to fail because of that:

https://ci.theforeman.org/blue/organizations/jenkins/foreman-installer-develop-package-release/detail/foreman-installer-develop-package-release/1538/pipeline/176

**Review by ekohl — APPROVED** [review]

**Pull request merged**